### PR TITLE
[RFR] allow to pass d3 as argument, rather than using a global instance

### PR DESCRIPTION
**Pull request by ThieryMichel**
Allow to either use a global d3 (eg from a cdn) or an imported one from dependencies.

### CODE REVIEW

**Comment by fzaninotto** [comment]
It's a BC break, isn't it?

**Comment by ThieryMichel** [comment]
nope you can still use event drops with a global d3, if you want.
Also the rewrite branch already break backward compatibility.

File: demo/demo.js [hunk 1]
```diff
@@ -5,8 +5,7 @@ import '../src/style.css';
 
 const repositories = require('./data.json');
 
-const chart = eventDrops();
```
Comment (jpetitcolas) [inline demo/demo.js]: I don't like signature of this function. Forcing to pass an empty object to specify d3 looks weird. I would simply pass nothing here.

File: src/index.js [hunk 2]
```diff
@@ -31,12 +31,15 @@ export const draw = (config, xScale) =>
             .data(filteredData)
             .call(dropLine(config, xScale))
             .call(bounds(config, xScale))
-            .call(axis(config, xScale));
+            .call(axis(d3, config, xScale));
     };
 
-export default (customConfiguration = {}) =>
+export default (customConfiguration = {}, d3 = window.d3) =>
```
Comment (jpetitcolas) [inline src/index.js]: s/window/global/

File: src/index.js [hunk 3]
```diff
@@ -31,12 +31,15 @@ export const draw = (config, xScale) =>
...
     selection => {
-        const config = defaultsDeep(customConfiguration, defaultConfiguration);
+        const config = defaultsDeep(
+            customConfiguration,
```
Comment (jpetitcolas) [inline src/index.js]: Won't it be any issues with a custom D3 used here in comparison of the `d3` used in EventDrops?

File: src/zoom.js [hunk 4]
```diff
@@ -1,10 +1,10 @@
-export default (svg, config, xScale, draw) => {
+export default (d3, svg, config, xScale, draw, getEvent) => {
     const zoom = d3.zoom();
 
     zoom.on('zoom', () => {
-        const newScale = d3.event.transform.rescaleX(xScale);
+        const newScale = getEvent().transform.rescaleX(xScale);
```
Comment (jpetitcolas) [inline src/zoom.js]: Can't we use `import {event as currentEvent} from 'd3';` instead?

File: src/index.js [hunk 5]
```diff
@@ -53,6 +56,8 @@ export default (customConfiguration = {}) =>
             margin,
         } = config;
 
+        const getEvent = () => d3.event; // keep d3.event mutable see https://github.com/d3/d3/issues/2733
```
Comment (jpetitcolas) [inline src/index.js]: I prefer the way enounced here: https://github.com/d3/d3/issues/2733#issuecomment-190743489

File: src/zoom.js (hunk 4 above)
Comment (ThieryMichel) [reply src/zoom.js]: If we do that we lose the ability to use global d3.

File: src/index.js (hunk 5 above)
Comment (ThieryMichel) [reply src/index.js]: will not work with a global d3

File: src/index.js (hunk 2 above)
Comment (ThieryMichel) [reply src/index.js]: global will only work if using webpack no?

**Pull request merged**